### ISSUE TITLE
Edit Free Listings: Success notification is shown when there is an API error and saving is not successful

### ISSUE THREAD

**Issue by ecgan** [woocommerce/google-listings-and-ads#1331]
### Describe the bug:

- This is mentioned in https://github.com/woocommerce/google-listings-and-ads/pull/1313#pullrequestreview-911315044.

> Found another issue that already existed before this project. When there is a failed API call, the notification at the bottom still shows "Your changes to your Free Listings have been saved and will be synced to your Google Merchant Center account."

https://github.com/woocommerce/google-listings-and-ads/blob/154565b1739ba5a9982caaab60447a05503d00ba/js/src/edit-free-campaign/index.js#L188-L208

https://user-images.githubusercontent.com/17420811/158565390-9cf4b921-0adf-4eda-b37e-a6ec65f972a2.mp4

### Steps to reproduce:

1. Modify some API path (e.g. upsert shipping rate API path) so that it will throw error when users are trying to save changes in Edit Free Listings.
2. Go to Edit Free Listings and make some changes.
3. Click on "Save changes" button.

### Expected behavior:

Error message notification should be displayed, e.g. `'Something went wrong while saving your changes. Please try again later.'`.

### Actual behavior:

Success message notification is displayed: `'Your changes to your Free Listings have been saved and will be synced to your Google Merchant Center account.'`.

### Additional details:

I think this is due to the wrong use of [`Promise.allSettled()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled). It should be [`Promise.all()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all) instead. We should investigate a lil bit deeper on why we used `allSettled` in the first place, whether it is really intentional.

**Comment by jorgemd24** [comment]
I was having a look at this issue and I have two questions:

1. I understand that the use of [Promise.allSettled()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled) is because we allow to sync the data even when it is partially successful but is it OK to sync the data even if the rates are not saved successful? Should not it be [Promise.all()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all) ? I am tagging @eason9487 😄   as I see that you reviewed this part of the code before https://github.com/woocommerce/google-listings-and-ads/pull/468#discussion_r613085411

https://github.com/woocommerce/google-listings-and-ads/blob/8e0ae939a647850e74babe1cb4d1497dee1b9be3/js/src/edit-free-campaign/index.js#L199-L200

2. Related to the array of promises that we are waiting to be fulfilled:

https://github.com/woocommerce/google-listings-and-ads/blob/8e0ae939a647850e74babe1cb4d1497dee1b9be3/js/src/edit-free-campaign/index.js#L192-L197

I see that `saveShippingRates` -> `upsertShippingRates` , does not have any try and catch, so in this case, if the upserShippingRates fails, it will fail silently, is this intentional? I am tagging @ecgan as I see that you did some work here 😄 

https://github.com/woocommerce/google-listings-and-ads/blob/249de237fe21bcb4f757b13e6d2d2d4ccafdb2e2/js/src/data/actions.js#L103-L123

**Comment by eason9487** [comment]
> I understand that the use of [Promise.allSettled()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled) is because we allow to sync the data even when it is partially successful but is it OK to sync the data even if the rates are not saved successful? Should not it be [Promise.all()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all) ?

When reviewing the revision of #468, I guess we only focused on the design problem of the silent error, but didn't confirm whether it's OK to sync the data when error occurred. So this might needs more survey. I guess Ideally, if there are any errors that are catchable, then it should be better not to call the `fetchSettingsSync()`.

Apart from the "allowing to call sync the data (`fetchSettingsSync()`)", the intent of using `Promise.allSettled()` to "save all other data partially" is right. it will give us more controllable options when we want to handle errors thoroughly.

If we scope this issue on the frontend parts with the current implementation, then I think  `Promise.allSettled()` is the correct use here. Suppose we had implemented [the proposed error handling](https://github.com/woocommerce/google-listings-and-ads/issues/489) or all these *wp-data* `actions` were able to throw out errors, it would be nice to wait for all API calls to be finished and then pick out the failed calls to determine the next action. For example:

- We could add a retry feature if the error falls into a resumable reason. And if the retry is successful, we don't even need to let merchants know the error ever happened because it's smoother.
- We could show more concrete messages to merchants if multiple errors occurred.

---

In my assumption, the most ideal way is to implement the transaction mechanism for these data mutations so that the partially updated data can be rollbacked once any API calls inside `Promise.allSettled()` fails. But this is the perspective of data layer, and probably it requires a separate issue.

**Comment by ecgan** [comment]
@jorgemd24 ,

> I see that `saveShippingRates` -> `upsertShippingRates` , **_does not have any try and catch_**, so in this case, **_if the upserShippingRates fails, it will fail silently_**, is this **_intentional_**? I am tagging @ecgan as I see that you did some work here 😄

The `saveShippingRates` -> `upsertShippingRates` does not have try/catch, so when there is an error, the error will be thrown out and can be caught by the consumer. It does not fail silently (I'm guessing you had a typo? Or do we have different interpretation here? 🤔  ). This is intentional, it was done in PR https://github.com/woocommerce/google-listings-and-ads/pull/1313. The idea is similar to issue https://github.com/woocommerce/google-listings-and-ads/issues/489, so that the consumer is able to know whether the API calls are successful or not.

---

@eason9487 ,

> If we scope this issue on the frontend parts with the current implementation, then I think `Promise.allSettled()` is the correct use here. Suppose we had implemented [the proposed error handling](https://github.com/woocommerce/google-listings-and-ads/issues/489) or all these _wp-data_ `actions` were able to throw out errors, it would be nice to wait for all API calls to be finished and then pick out the failed calls to determine the next action. For example: [...]

That is an interesting idea 🙂 👍 I think that is a nice-to-have enhancement that we can have one day as a separate issue. However, for this high-priority bug issue, I guess we can just change `Promise.allSettled` to `Promise.all` and get it fixed? (i.e. like you mentioned, "if there are any errors that are catchable, then it should be better not to call the `fetchSettingsSync()`".)

**Comment by jorgemd24** [comment]
@ecgan,

>  It does not fail silently (I'm guessing you had a typo? Or do we have different interpretation here? 🤔 )

In the current scenario, with the following promises:

https://github.com/woocommerce/google-listings-and-ads/blob/8e0ae939a647850e74babe1cb4d1497dee1b9be3/js/src/edit-free-campaign/index.js#L192-L197

I see that if `saveTargetAudience` and `saveSettings` fail I am able to see a prompt in the front-end, this is happening because the try and catch are in the action itself. 

![image](https://user-images.githubusercontent.com/2488994/164311305-caa0f84e-b303-4b7d-a81e-498ad9dac2b9.png)

https://github.com/woocommerce/google-listings-and-ads/blob/71c415c8513d8a086f149e605b60a13d65b27465/js/src/data/actions.js#L720-L741

On the other hand, if `saveShippingRates`and `saveShippingTimes` fail,  they are going to fail silently in this case because the consumer is not catching the results from the `Promise.allSettled()`.

So my question was more related why `saveTargetAudience` and `saveSettings` were using the try and catch in the action itself and `saveShippingRates`and `saveShippingTimes` were not but now I see the reason here:

> No try/catch block in upsertShippingRates and deleteShippingRates wp.data actions and in saveShippingRates custom hook, so that the error can propagate up and be caught by the components to display the necessary error message based on the context.

So I understand that the idea is to leave the consumer to handle the exeptions instead of the action, is this right?

**Comment by eason9487** [comment]
@ecgan

> That is an interesting idea 🙂 👍 I think that is a nice-to-have enhancement that we can have one day as a separate issue.

It's not just an idea. It was the original requirement but limited by the implementation of **actions.js**. Quoted from https://github.com/woocommerce/google-listings-and-ads/pull/468#issuecomment-819179335, the proposed error message was:

> For handling multiple save requests, let's change the copy to: `There was an error trying to [do some action]. Other changes have been saved. Please try again later.`

Assume the error message based on the original proposal to interpret the error message when multiple errors occur. I think it's reasonable to show something like `There are errors trying to [do A action and B action]. Other changes have been saved. Please try again later.`. The handling will be implemented with `Promise.allSettled()` eventually.

Therefore, in order to retain this goal and to be compatible with future improvements, I would suggest adding the further processing for picking out the errors from the current `Promise.allSettled()` results instead of changing it back and forth to `Promise.all()` and then `Promise.allSettled()`. This will take a little more time than changing it to `Promise.all()`, but it's not much and should be worth it.

**Comment by ecgan** [comment]
@jorgemd24 ,

> On the other hand, if `saveShippingRates`and `saveShippingTimes` fail, they are going to fail silently in this case because the consumer is not catching the results from the `Promise.allSettled()`.

Ahh okay, now I get your meaning of "fail silently" (I thought it was a try/catch block, catching error and do nothing). Thanks. 🙂 

> So I understand that the idea is to leave the consumer to handle the exeptions instead of the action, is this right?

Yes, correct.

---

@eason9487 ,

> It's not just an idea. It was the original requirement but limited by the implementation of **actions.js**. Quoted from [#468 (comment)](https://github.com/woocommerce/google-listings-and-ads/pull/468#issuecomment-819179335), the proposed error message was:
> 
> > For handling multiple save requests, let's change the copy to: `There was an error trying to [do some action]. Other changes have been saved. Please try again later.`

Ahh gotcha, thanks for the link and quote. 🙂 Yeah, to do that, it would need to be implemented with `Promise.allSettled()` eventually.

> Assume the error message based on the original proposal to interpret the error message when multiple errors occur. I think it's reasonable to show something like `There are errors trying to [do A action and B action]. Other changes have been saved. Please try again later.`.

I'm thinking about the implementation detail on doing this. Putting a note here. It may not be as straightforward as it seems, especially with i18n.

Within the `Promise.allSettled()` call, we have four concurrent operations. Depending on the number of errors from there, we would need to have different error message strings, e.g.

- One error: `There are errors trying to [do A action]. Other changes have been saved. Please try again later.`
- Two errors: `There are errors trying to [do A action] and [do B action]. Other changes have been saved. Please try again later.` (Notice the "and".)
- Three errors: `There are errors trying to [do A action], [do B action] and [do C action]. Other changes have been saved. Please try again later.` (Notice the comma "," and "and". )
- All four errors: `There are errors trying to [do all actions]. Please try again later.` (No mention of "other changes", since all failed.)

What this means is that I think we would need to have four i18n strings to cater for the above situations. I'm not sure if there are better ways to produce the strings. 🤔 

Here's a related link I found on StackOverflow: [l10n/i18n: how to handle phrases with dynamic list of items?](https://stackoverflow.com/q/45802189/379641)

Alternatively, instead of using one paragraph like the above, maybe we can consider using bullet points to show the error messages, e.g.:

```
There are errors in the following actions: 

- Saving your shipping rates.
- Saving your shipping times.

Other changes have been saved. Please try again later.
```

This is provided if we can use bullet points in the notification component. I think this is a bit more straightforward, and also scalable to any number of errors.

**Comment by jorgemd24** [comment]
Thanks, @eason9487 and @ecgan for the comments 🙏 

I think now it is clear that using `Promise.allSettled` provides some benefits in this scenario. However, I have my doubts if it is OK to sync the data to the MC when saveShippingRates, saveShippingTimes, etc are failing.  As @eason9487 mentioned before, we will need to do more surveys about it.

From the front-end perspective, I see the benefits to have more control over the errors caused in the API calls, so I think we could: 

- Continue with the implementation mentioned here: https://github.com/woocommerce/google-listings-and-ads/issues/438#issuecomment-821093961 or discuss other alternatives.
- Once that part is done, we could retrieve the errors (if any) of the following promises:
https://github.com/woocommerce/google-listings-and-ads/blob/8e0ae939a647850e74babe1cb4d1497dee1b9be3/js/src/edit-free-campaign/index.js#L192-L197
- If there are errors, we can list the errors as @ecgan  mentioned before, avoiding the problems related to i18n strings:

>Alternatively, instead of using one paragraph like the above, maybe we can consider using bullet points to show the error messages, e.g.:

>There are errors in the following actions: 
> - Saving your shipping rates.
> - Saving your shipping times.

>Other changes have been saved. Please try again later.

What are your opinions @ecgan & @eason9487?

**Comment by eason9487** [comment]
> * Continue with the implementation mentioned here: [Default shipping state, wp-data's `isResolving`, vs. `Form.Props.initialValues`  #438 (comment)](https://github.com/woocommerce/google-listings-and-ads/issues/438#issuecomment-821093961) or discuss other alternatives.

The approach I mentioned in that comment focuses on the error state access of API calls triggered by wp-data selectors and will require a larger code rewrite. While the same pattern could be applied to wp-data actions, I think the "[Let actions signal error explicitly](https://github.com/woocommerce/google-listings-and-ads/issues/489#issuecomment-820598091)" mentioned by @tomalec is also needed in actions, which can make the code consumer side of actions smoother and more flexible.

If considering this issue's scope, a compatible transitional approach could be to change these needed actions to throw errors, and add a new custom react hook to generically display error messages to replace the `handleFetchError` function for calls that don't require special error handling.

**Comment by ecgan** [comment]
@jorgemd24 , yeah, I agree with @eason9487's comment above too.

To be more specific, this is what I'm thinking:

```js
const results = await Promise.allSettled( [
	saveTargetAudience( targetAudience ), // TODO: remove its try/catch in actions.js.
	saveSettings( settings ), // TODO: remove its try/catch in actions.js.
	saveShippingRates( shippingRates ), // already throws error when it fails.
	saveShippingTimes( shippingTimes ), // already throws error when it fails.
] );
```

With the above, we can get access to the success and error results in `results`. One thing to note: we need to be careful with the other impacted areas when we "remove its try/catch in actions.js" for `saveTargetAudience` and `saveSettings`, I think it might be good to split the changes into different PRs.

**Comment by jorgemd24** [comment]
Thanks, @ecgan,

Yeah, that makes sense, I will start from there and I think once that part is done we can add a new custom hook to display error messages as @eason9487 mentioned before.
